src/server/ua_server_binary.c

```diff
@@ -996,3 +997,206 @@ UA_Server_networkCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
+                        "Attempt to reverse reconnect to %.*s:%d", (int)rc->hostname.length,
+                        rc->hostname.data, rc->port);
+            UA_Server_attemptReverseConnect(server, rc);
+        }
+    }
+}
+
+UA_StatusCode UA_Server_setReverseConnectRetryCallback(UA_Server *server, UA_Boolean enabled) {
+    if (enabled && !server->reverseConnectsCheckHandle) {
+        const UA_UInt32 reconnectInterval = server->config.reverseReconnectInterval ?
+                    server->config.reverseReconnectInterval : 15000;
+        return UA_Server_addRepeatedCallback(server, retryReverseConnectCallback, NULL,
+                                             reconnectInterval, &server->reverseConnectsCheckHandle);
+    } else if (!enabled && server->reverseConnectsCheckHandle) {
+        UA_Server_removeCallback(server, server->reverseConnectsCheckHandle);
+        server->reverseConnectsCheckHandle = 0;
+        return UA_STATUSCODE_GOOD;
+    }
+
+    return UA_STATUSCODE_GOOD;
+}
+
+static void setReverseConnectState(UA_Server *server, reverse_connect_context *context,
+                                   UA_SecureChannelState newState) {
+    if (context->state == newState)
+        return;
+
+    context->state = newState;
+
+    if (context->stateCallback)
+        context->stateCallback(server, context->handle, context->state,
+                               context->callbackContext);
+}
+
+void UA_Server_reverseConnectCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
+                          void *application, void **connectionContext,
+                          UA_ConnectionState state, const UA_KeyValueMap *params,
+                          UA_ByteString msg) {
+    (void)params;
+
+    UA_Server *server = (UA_Server*)application;
+
+    /* UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER, "Reverse connect has state %d", state); */
+
+    reverse_connect_context *context = (reverse_connect_context *)*connectionContext;
+
+    context->currentConnection.state = state;
+
+    if(context->currentConnection.connectionId == 0) {
+        /* The socket is closing without being previously registered -> ignore */
+        if(state == UA_CONNECTIONSTATE_CLOSED ||
+           state == UA_CONNECTIONSTATE_CLOSING) {
+            setReverseConnectState(server, context, UA_SECURECHANNELSTATE_CONNECTING);
+            return;
+        }
+
+        context->currentConnection.state = state;
+        context->currentConnection.connectionId = connectionId;
+        context->currentConnection.connectionManager = cm;
+        return;
+    }
+
+    /* The connection is closing. This is the last callback for it. */
+    if(state == UA_CONNECTIONSTATE_CLOSING && context->channel) {
+        deleteServerSecureChannel(server, context->channel);
+        context->channel = NULL;
+        setReverseConnectState(server, context, UA_SECURECHANNELSTATE_CONNECTING);
+        context->currentConnection.connectionId = 0;
```

Comment:
If the connection is lost, try to establish a new connection immediately.
So don’t wait for the interval.